src/pykeen/evaluation/rank_based_evaluator.py

```diff
@@ -64,6 +70,88 @@
 RANK_OPTIMISTIC, RANK_REALISTIC, RANK_PESSIMISTIC = RANK_TYPES
 
 
+@dataclass
+class ValueRange:
+    """A value range description."""
+
+    #: the lower bound
+    lower: Optional[float] = None
+
+    #: whether the lower bound is inclusive
+    lower_inclusive: bool = False
+
+    #: the upper bound
+    upper: Optional[float] = None
+
+    #: whether the upper bound is inclusive
+    upper_inclusive: bool = False
+
+    def __contains__(self, x: SupportsFloat) -> bool:
+        if self.lower is not None:
+            if x < self.lower:
+                return False
+            if not self.lower_inclusive and x == self.lower:
+                return False
+        if self.upper is not None:
+            if x > self.upper:
+                return False
+            if not self.upper_inclusive and x == self.upper:
+                return False
+        return True
+
+
+class RankBasedMetric:
```

Comment:
@cthoyt what do you think about this?

Reply:
I like the idea, using the class allows us to annotate a lot of useful information in a first-class way, then later use the class-resolver. Maybe you saw in #705 I also took advantage of a similar idea of annotating this information into RexMex (which I actually annotated by hand in PyKEEN first)

Reply:
Overall I think a lot of the improvements to use classes instead of ad-hoc data structures made both the docs and code cleaner. Nice job